### PR TITLE
Fix affine data copy generation corner cases/bugs

### PR DESCRIPTION
**Pull request by bondhugula**
- the [begin, end) range identified for copying could end in between the
  block, which makes hoisting invalid in some cases. Change the range
  identification to always end with end of block.

- add test case to exercise these (with fast mem capacity set to minimal so
  that single element memref buffers are generated at the innermost loop)

- the location of begin/end of the block range for data copying was
  being confused with the insert points for copy in and copy out code.
  In cases, where we choose to hoist transfers, these are separate.

- when copy loops are single iteration ones, promote their bodies at
  the end of the pass.

- change default fast mem space to 1 (setting it to zero made it
  generate DMA op's that won't verify in the default case - since the
  DMA ops have a check for src/dest memref spaces being different).

Signed-off-by: Uday Bondhugula <uday@polymagelabs.com>

### CODE REVIEW

**Comment by antiagainst** [comment]
Hey Uday, sorry for the delay. It seems this PR needs a rebase now.

**Comment by antiagainst** [comment]
@nicolasvasilache could you review this?

File: lib/Transforms/AffineDataCopyGeneration.cpp [hunk 1]
```diff
@@ -625,35 +639,48 @@ LogicalResult AffineDataCopyGeneration::runOnBlock(Block *block) {
...
+      // Get to the start of the next range.
+      curBegin = std::find_if(std::next(it), block->end(), [&](Operation &op) {
+        return (isa<AffineLoadOp>(op) || isa<AffineStoreOp>(op) ||
+                isa<AffineForOp>(op)) &&
+               copyNests.count(&op) == 0;
+      });
+      it = curBegin;
+    } else {
+      // We simply include this op in the current range and continue for more.
+      ++it;
```
Comment (andydavis1) [inline lib/Transforms/AffineDataCopyGeneration.cpp]: These changes look fine, but this function (and loop here) is getting quite long. It would be nice to have a follow up PR that broken this pass up into functions and refactored things a bit...

Comment (bondhugula) [reply lib/Transforms/AffineDataCopyGeneration.cpp]: Yes, I'll do that next. I actually have some changes ready that refactor this pass code into a separate data copy generation utility moved into lib/Transforms/Utils/Utils.cpp - so that it's reusable elsewhere. I'm waiting for this to get in.

**Comment by bondhugula** [comment]
> Hey Uday, sorry for the delay. It seems this PR needs a rebase now.

Done. Do you want me to rebase, squash and force push?

**Comment by antiagainst** [comment]
@bondhugula: yes please squash into one commit and force push. Thanks! :)

**Comment by bondhugula** [comment]
> @bondhugula: yes please squash into one commit and force push. Thanks! :)

Done.

File: lib/Transforms/AffineDataCopyGeneration.cpp (hunk 1 above)
Comment (bondhugula) [reply lib/Transforms/AffineDataCopyGeneration.cpp]: I've now turned this into a while loop (didn't make a lot of sense as a 'for'). This loop is a one-pass on a block of op's and it's processing (contiguous) ranges of op's from among it - while skipping over portions. So it's really not a structured one. It's now shorter with one of the cases eliminated - I'll see if this could be made more readable in the next change on this.

File: lib/Transforms/AffineDataCopyGeneration.cpp [hunk 2]
```diff
@@ -625,35 +639,48 @@ LogicalResult AffineDataCopyGeneration::runOnBlock(Block *block) {
...
+                isa<AffineForOp>(op)) &&
+               copyNests.count(&op) == 0;
+      });
+      it = curBegin;
+    } else if (copyNests.count(&*it) > 0) {
+      // This op is a copy nest and the range ends here.
       runOnBlock(/*begin=*/curBegin, /*end=*/it);
-      curBegin = std::next(it);
+      // Get to the start of the next range.
+      curBegin = std::find_if(std::next(it), block->end(), [&](Operation &op) {
```
[the remark this answers is not in the document]
Comment (bondhugula) [reply lib/Transforms/AffineDataCopyGeneration.cpp]: Realized that this is no longer reached due to the check  'copyNests.count(&op) == 0; ' while looking for the next iteration point. Removed this, and added an assertion.

File: lib/Transforms/AffineDataCopyGeneration.cpp [hunk 3]
```diff
@@ -625,35 +639,48 @@ LogicalResult AffineDataCopyGeneration::runOnBlock(Block *block) {
...
+        return (isa<AffineLoadOp>(op) || isa<AffineStoreOp>(op) ||
+                isa<AffineForOp>(op)) &&
+               copyNests.count(&op) == 0;
```
[the remark this answers is not in the document]
Comment (bondhugula) [reply lib/Transforms/AffineDataCopyGeneration.cpp]: Realized that this is no longer reached due to the check  'copyNests.count(&op) == 0; ' while looking for the next iteration point. Removed this, and added an assertion. Thanks. 
Coverage should be fine now.

**Comment by bondhugula** [comment]
> All (the pull request submitter and all commit authors) CLAs are signed, **but** one or more commits were authored or co-authored by someone other than the pull request submitter.
> 
> We need to confirm that all authors are ok with their commits being contributed to this project. Please have them confirm that by leaving a comment that contains only `@googlebot I consent.` in this pull request.
> 
> _Note to project maintainer:_ There may be cases where the author cannot leave a comment, or the comment is not properly detected as consent. In those cases, you can manually confirm consent of the commit author(s), and set the `cla` label to `yes` (if enabled on your project).
> 
>  **Googlers: [Go here](https://goto.google.com/prinfo/https%3A%2F%2Fgithub.com%2Ftensorflow%2Fmlir%2Fpull%2F88) for more info**.



> All (the pull request submitter and all commit authors) CLAs are signed, **but** one or more commits were authored or co-authored by someone other than the pull request submitter.
> 
> We need to confirm that all authors are ok with their commits being contributed to this project. Please have them confirm that by leaving a comment that contains only `@googlebot I consent.` in this pull request.
> 
> _Note to project maintainer:_ There may be cases where the author cannot leave a comment, or the comment is not properly detected as consent. In those cases, you can manually confirm consent of the commit author(s), and set the `cla` label to `yes` (if enabled on your project).
> 
>  **Googlers: [Go here](https://goto.google.com/prinfo/https%3A%2F%2Fgithub.com%2Ftensorflow%2Fmlir%2Fpull%2F88) for more info**.

@joker-eph You'll have to do a '@googlebot I consent' or if you prefer I can exclude you from the commit message (this was for the formatting fix suggestion).

File: lib/Transforms/AffineDataCopyGeneration.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (antiagainst) [reply lib/Transforms/AffineDataCopyGeneration.cpp]: @River707: could you check again and approve if okay?

File: test/Transforms/affine-data-copy.mlir [hunk 4]
```diff
@@ -107,3 +109,62 @@ func @matmul(%A: memref<4096x4096xf32>, %B: memref<4096x4096xf32>, %C: memref<40
...
+//
+<<<<<<< HEAD
+// This test case will lead to single element buffers. These are eventually
+// expected to be turned into registers via alloca and mem2reg.
+=======
+// This test case will minimal fast mem capacity will lead to single element
+// memref buffers. These are eventually expected to be turned into registers via
+// alloca and mem2reg as the single iteration loops around them are cleaned
+// up.
+>>>>>>> 95a26422... Fix affine data copy generation corner cases/bugs
```
Comment (River707) [inline test/Transforms/affine-data-copy.mlir]: Looks like there are some merge conflicts in here.

Comment (bondhugula) [reply test/Transforms/affine-data-copy.mlir]: Sorry about this - fixed.

**Comment by bondhugula** [comment]
> The walkers were just updated to remove the need for the explicit <>.

Merged and fixed.

**Comment by bondhugula** [comment]
@joker-eph I think this requires your author consent again.
